### PR TITLE
add pathof(::Module)

### LINKED ISSUE #27592: Expose Base.find_package?
`Base.find_package` is similar to `Pkg.dir` except that it reports the actual path that would be loaded instead of just joinpathing things.

```
julia> Base.find_package("UUIDs")
"/Users/kristoffer/julia/usr/share/julia/stdlib/v0.7/UUIDs/src/UUIDs.jl"

julia> Base.find_package("SpecialFunctions")
"/Users/kristoffer/.julia/v0.6/SpecialFunctions/src/SpecialFunctions.jl"

julia> Base.find_package("Example")
"/Users/kristoffer/.julia/packages/Example/kH44/src/Example.jl"
```

Could be useful to expose.

### PR DESCRIPTION
**Pull request by stevengj**
Closes #27592.   `abspath(FooModule)` returns the path of `FooModule.jl`.    The package directory can be obtained from `abspath(FooModule, "..", "..")`.

### CODE REVIEW

**Comment by fredrikekre** [comment]
Should this really be a `Pkg` function? Isn't this useful for other modules as well?

**Comment by stevengj** [comment]
@fredrikekre, it doesn't work for arbitrary modules, since it only makes sense for modules that come from a standard `Pkg` directory structure.  (Modules don't even have to come from files!)

**Comment by KristofferC** [comment]
True, but the answer to the question from a package will be loaded has nothing to do with the implementation of the package manager. Forcing a dependency on Pkg for something that it just directly asks Base code loading about seems unfortunate?

**Comment by stevengj** [comment]
Sure, it's easy to move this to base.  What should the function be called, and should it be exported?

**Comment by stevengj** [comment]
How about `abspath(::Module)`?

**Comment by ararslan** [comment]
That feels like a little bit of a pun but I kind of like it.

**Comment by fredrikekre** [comment]
> How about `abspath(::Module)`?

I would expect this to return `/path/to/package/src/Package.jl` and not `/path/to/package/`. Maybe returning `/path/to/package/src/Package.jl` is more useful anyway?

**Comment by StefanKarpinski** [comment]
Maybe every module should just know what source file it was defined in with `nothing` being the answer if it was defined in the REPL or some other non-source-file location?

**Comment by KristofferC** [comment]
> Yes I believe so, especially because AFAIK there's nothing that says a package has to have a particular directory structure,

Yes, there is, look for "src" in `loading.jl`.

**Comment by iamed2** [comment]
> Maybe every module should just know what source file it was defined in with nothing being the answer if it was defined in the REPL or some other non-source-file location?

Implicit `const __FILE__ = @__FILE__` in each module?

**Comment by ararslan** [comment]
> Yes, there is, look for "src" in `loading.jl`.

Oh. 😐 Alright then. Well, I still think returning the file is useful. 😛

**Comment by stevengj** [comment]
Changed to `abspath` and made to return the file.

**Comment by stevengj** [comment]
CI is looking good.

**Comment by stevengj** [comment]
Marking as triage as otherwise there is no non-deprecated way of getting this info without mucking around in Base internals.

**Comment by JeffBezanson** [comment]
@StefanKarpinski and/or @KristofferC will have to decide this one.

**Comment by StefanKarpinski** [comment]
I'm a little uneasy about overloading `abspath` for this but I supposed it seems ok. I'm also not super into the multiargument version since this seems like a somewhat more useful definition:
```jl
abspath(m, paths...) = joinpath(dirname(abspath(m)), paths...)
```

File: stdlib/Pkg/src/API.jl [hunk 1]
```diff
@@ -430,7 +430,7 @@ function clone(url::String, name::String = "")
 end
 
 function dir(pkg::String, paths::AbstractString...)
-    @warn "`Pkg.dir(pkgname, paths...)` is deprecated; instead, do `import $pkg; abspath($pkg, \"..\", \"..\", paths...)`." maxlog=1
+    @warn "`Pkg.dir(pkgname, paths...)` is deprecated; instead, do `import $pkg; joinpath(dirname(abspath($pkg)), \"..\", paths...)`." maxlog=1
```
Comment (StefanKarpinski) [inline stdlib/Pkg/src/API.jl]: With the `dirname` call you don't want the `..` anymore.

Comment (stevengj) [reply stdlib/Pkg/src/API.jl]: You still need `..`, because `dirname` is the `src` directory, and `Pkg.dir` returns the parent directory.

Comment (StefanKarpinski) [reply stdlib/Pkg/src/API.jl]: Ah, gotcha.

**Comment by ararslan** [comment]
> I'm a little uneasy about overloading `abspath` for this but I suppose it seems ok

It does feel like a bit of a pun. Packages are loaded based on `LOAD_PATH`; perhaps to get the path from which a package is loaded, one should use `loadpath(::Module)`?

**Comment by stevengj** [comment]
`which(module)`?

**Comment by StefanKarpinski** [comment]
Maybe `where(m)` instead?

**Comment by ararslan** [comment]
I don't particularly like `where`, both because we already use that as a (non-reserved) keyword and I don't find it very descriptive.

**Comment by ararslan** [comment]
Having `path` somewhere in the name of the function seems worthwhile, as it signifies that what you're asking for and getting is a path.

**Comment by stevengj** [comment]
A nice thing about `pathof` is that I can imagine it being extended to other things, e.g. file descriptors, IOStreams, etcetera.  (But we could also use `abspath` for this.)

**Comment by ararslan** [comment]
I think `pathto` reads a little nicer than `pathof`, since in prose you'd say "the path _to_ X," but I'm also fine with `abspath`.

**Comment by StefanKarpinski** [comment]
I don't like `pathto` because this is not the path to a module—whatever that is; it's the path of the file in which the module was defined, which `pathof` seems to pithily describe. Moreover, it's reminiscent of `nameof` which is a similar kind of reflection function.

**Comment by JeffBezanson** [comment]
Isn't this a new feature, and thus can be designed after 1.0?

**Comment by stevengj** [comment]
@JeffBezanson, it is and it isn't — the problem is that an old feature, `Pkg.dir`, was deprecated in 0.7 and there is no documented replacement.  So if we wait until 1.1 for this we will end up with a bunch of packages relying on undocumented Base internals.

**Comment by ararslan** [comment]
Yes, but it makes things like documentation deployment and coverage submission much easier in CI, since currently scripts have to call `Pkg.dir`. With this they would be able to do `julia -e 'using Package; include(joinpath(pathof(Package), ...); dostuff'`.

**Comment by stevengj** [comment]
Changed to `pathof`.

**Comment by fredrikekre** [comment]
> Yes, but it makes things like documentation deployment and coverage submission much easier in CI, since currently scripts have to call Pkg.dir. With this they would be able to do julia -e 'using Package; include(joinpath(pathof(Package), ...); dostuff'.

For this particular case you shouldn't need `Pkg.dir`-like functionality though, see e.g. https://github.com/JuliaDiff/ForwardDiff.jl/blob/d7ebc3d2eca3c1c4b10815cff8900bee9a600761/.travis.yml#L18

**Comment by ararslan** [comment]
CI is green and we're waiting on a couple of other things before tagging RC1, and with any luck that will be tagged before this week's triage. So can we make a decision here in short order?

**Pull request merged**